### PR TITLE
Add more blacklists to codegen

### PR DESCRIPTION
**Pull request by hausdorff**

### CODE REVIEW

**Comment by hausdorff** [comment]
We should probably collapse the beta patches down before checkin because they're each about 0.7 megabytes. Don't want to make the .git folder too much bigger than it has to be.

File: ksonnet-gen/kubeversion/data.go [hunk 1]
```diff
@@ -30,7 +30,104 @@ var versions = map[string]versionData{
...
+			"io.k8s.kubernetes.pkg.apis.rbac.v1alpha1.RoleList":                             newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.rbac.v1beta1.ClusterRoleBindingList":                newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.rbac.v1beta1.ClusterRoleList":                       newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.rbac.v1beta1.RoleBindingList":                       newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.rbac.v1beta1.RoleList":                              newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.settings.v1alpha1.PodPresetList":                    newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.storage.v1.StorageClassList":                        newPropertySet("metadata"),
+			"io.k8s.kubernetes.pkg.apis.storage.v1beta1.StorageClassList":                   newPropertySet("metadata"),
+
+			// Status fields.
```
Comment (anguslees) [inline ksonnet-gen/kubeversion/data.go]: :+1: I was going to suggest that any top-level "status" field should be added to this blacklist.

File: ksonnet-gen/ksonnet/emit.go [hunk 2]
```diff
@@ -422,6 +429,20 @@ func (ao *apiObject) emitAsRefMixins(
 	m.writeLine("},")
 }
 
+func (ao *apiObject) emitConstructor(m *indentWriter) {
+	if dm, ok := ao.propertyMethods["default"]; ok {
```
Comment (jbeda) [inline ksonnet-gen/ksonnet/emit.go]: I'm still not a huge fan of `default` as the constructor name over something like `new` or `init`.  But we can talk about that later.

File: ksonnet-gen/ksonnet/emit.go [hunk 3]
```diff
@@ -422,6 +429,20 @@ func (ao *apiObject) emitAsRefMixins(
...
+		log.Fatalf(
+			"Attempted to create constructor, but 'default' property already existed at '%s'",
+			dm.path)
+	}
+
+	if _, ok := ao.propertyMethods["metadata"]; !ok {
+		m.writeLine("default():: apiVersion + kind + {},")
+	} else {
+		m.writeLine("default(name):: apiVersion + kind + self.mixin.metadata.name(name),")
```
Comment (jbeda) [inline ksonnet-gen/ksonnet/emit.go]: There are a couple of problems with this:
1) Users may want to use `generateName`.  in that case, `name` must be blank.
2) It seems strange to have the name specified on things that often don't have names like "List" objects.  If we black out the metadata mixins we should probably black out name.

Comment (hausdorff) [inline ksonnet-gen/ksonnet/emit.go]: Good to know about `generateName`. I had no idea. That's a good enough reason to not take `name` as a default parameter.

But, just btw, regarding the second point, we're not blacking out all `metadata` mixins, we're only blacking out `metadata` mixins when they are referencing `ListMeta` (as opposed to `ObjectMeta`). So the effect is that anything with a `metadata` property that references `ObjectMeta` will end up getting a `name` property. As I said, given `generateName`, we still have to get rid of name, but I just wanted to be clear about what this is generating.

Comment (jbeda) [inline ksonnet-gen/ksonnet/emit.go]: SGTM -- let me know if you want to merge now or later.  Feel free to self merge now.

Comment (hausdorff) [inline ksonnet-gen/ksonnet/emit.go]: I'll merge after the wedding, I think.

File: ksonnet-gen/ksonnet/emit.go (hunk 2 above)
Comment (hausdorff) [inline ksonnet-gen/ksonnet/emit.go]: To follow up, we'll address this in the next design review.

**Comment by hausdorff** [comment]
Alright, as an update, we have smashed the changes to `ksonnet.beta.2` together and pulled them from this commit, because:

1. Since each revision costs 0.7 megabytes of space in the .git folder right now, we want to avoid bloating while we can.
2. We don't want to confuse people with a new version just yet.

I'm merging now.

**Pull request merged**